### ISSUE TITLE
XSS & XML entity expansion using via svg file (XML rendering) in File uploads 

### ISSUE THREAD

**Issue by rishaldwivedi** [GetSimpleCMS/GetSimpleCMS#1292]
Step to reproduce - 

- Upload (http://localhost/GetSimpleCMS-master/admin/upload.php) the malicious svg file & browse though, **XSS** &  **XML entity expansion attack**( lead to DOS) would be reproduced.

![simplecms](https://user-images.githubusercontent.com/20737069/43161948-6a86dc0c-8f3e-11e8-87e4-1856ecd239be.png)

![getsimplecms 1](https://user-images.githubusercontent.com/20737069/43162117-e3c585a0-8f3e-11e8-9b83-28f23b6989ec.png)

![getsimplecms 2](https://user-images.githubusercontent.com/20737069/43162118-e40e4114-8f3e-11e8-9197-dc08c6b67a34.png)

![getsimplecms 3](https://user-images.githubusercontent.com/20737069/43162191-1d8c3bee-8f3f-11e8-8670-9d2c8a80a4d4.png)

**Comment by tablatronix** [comment]
Mitigation , maybe filter svg uploads for malicous content.

**Comment by rishaldwivedi** [comment]
All the best with that, not sure if blacklist approach would be fair enough.  SVG provides a lot of flexibility to an attacker. 

Below are the possible workarounds:
- Don’t allow SVG
- Allow SVG submissions but don’t display them, just allow them to be downloaded
- Strip out dangerous stuff from the SVG before displaying. (Be careful to catch everything)
- Convert to a different image format before displaying, eg PNG or JPEG.

Ref: https://www.grepular.com/Scalable_Vector_Graphics_and_XSS

**Comment by bigin** [comment]
Just my opinion, it has only a low potential vulnerability:

1. You need admin rights to upload SVG file. 
1. Embedding SVG via `<img>` "should" be harmless.

It should be clear to everyone in the meantime that SVG are not just pictures but small applications.

Apart from upload SVG there are other ways for admins to execute "potential" dangerous code:
eval()
create_function()
assert()
include()
require()
etc 
... or just another dumb coding in components or template editor, should these all options be forbidden now?

If you do not want that in the core, there should be at least one way to enable SVG support again, maybe a config variable with a note?

Personally, I see not really urgent need for action here, but that's just my opinion.

**Comment by tablatronix** [comment]
No but making some things default prohibited and requiring that the end user explicitly enables it and understands the risks might be useful. They can also implement their own protections as needed.

This is a good idea to remove svg from out whitelist.

**Comment by rishaldwivedi** [comment]
@bigin  Rightly pointed by you `it has only a low potential vulnerability`, as it will be requiring some quality social engineering to make the file be available on the server. 

- When talking about `template editor`, do you imagine fooling an admin to make code level changes by tricking them?.  Instead tricking them upload a file, which may seem legit to them (but secretly hiding some malicious payload inside) is still viable.  

- Coming to another scenario, what if someone goes forward to customize the application by adding more user groups to it?. 
Few having access to upload functionality, but restrictions to admin related functionalities. Critical right?.

These kind of attack scenarios never end. So my opinion comes in like getting rid of these kind of low vulnerabilities is what makes a secure application, giving no way for an attacker to get in. 

> Embedding SVG via <img> "should" be harmless.

Sounds incomplete!, at the end of the day the svg file still resides & can be accessed directly. 
To make it complete, force download the file when accessed directly.

**Comment by rishaldwivedi** [comment]
@tablatronix  Fair enough!

**Comment by tablatronix** [comment]
Force download how, thats a hosting mime configuration. 

I mean we can htaccess something, but that does not cover all hosts and can break stuff on hosts with strict override rules.

**Comment by bigin** [comment]
@rishaldwivedi 
>- When talking about template editor, do you imagine fooling an admin to make code level changes by >tricking them?. Instead tricking them upload a file, which may seem legit to them (but secretly hiding >some malicious payload inside) is still viable.
>
>- Coming to another scenario, what if someone goes forward to customize the application by adding >more user groups to it?.
>Few having access to upload functionality, but restrictions to admin related functionalities. Critical >right?

Agree, what about a simple url? A url looks even more harmless, if no upload is allowed, an SVG can still be added: 

`<embed type="image/svg+xml" src="https://foooo.com/image.svg" />`
or
`<iframe src="https://foooo.com/image.svg"></iframe>`

Tell me how can this be prevented if admin is a naive guy? ;-)

>These kind of attack scenarios never end. So my opinion comes in like getting rid of these kind of low >vulnerabilities is what makes a secure application, giving no way for an attacker to get in.

A secure application on the web is a fantasy and will never exist. It is important that you don't make the application boring by senseless restrictions. Trying to control everything can make you paranoid.

>Sounds incomplete!, at the end of the day the svg file still resides & can be accessed directly. 
>To make it complete, force download the file when accessed directly.

That's correct.

However, I agree with the restriction to remove svg from our whitelist.

**Comment by rishaldwivedi** [comment]
@tablatronix True, my bad.

**Comment by rishaldwivedi** [comment]
> A secure application on the web is a fantasy and will never exist.

+1 to this 🤣

**Comment by tablatronix** [comment]
I saw this and then went and read about svg injections and thought, shit might as well just turn the computers off...

😄
